datastore/redis.go

```diff
@@ -386,38 +393,83 @@ func (r *RedisCache) SaveLatestBuilderBid(slot uint64, builderPubkey, parentHash
+	keyBidValues := r.keyBlockBuilderLatestBidsValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
 	bidValueMap, err := r.client.HGetAll(context.Background(), keyBidValues).Result()
 	if err != nil {
-		return err
+		return wasUpdated, err
 	}
 
-	// Find bid with highest value among all the latest bids
-	topBidValue := big.NewInt(0)
-	topBidBuilderPubkey := ""
+	// 2. Find bid with highest value among all the latest bids
 	for builderPubkey, bidValue := range bidValueMap {
 		val := new(big.Int)
 		val.SetString(bidValue, 10)
-		if val.Cmp(topBidValue) > 0 {
-			topBidValue = val
-			topBidBuilderPubkey = builderPubkey
+		if val.Cmp(prevTopBidValue) > 0 {
+			prevTopBidValue = val
+			prevTopBidBuilderPubkey = builderPubkey
 		}
 	}
 
-	if topBidBuilderPubkey == "" {
-		return ErrFailedUpdatingTopBidNoBids
+	// 3. Do nothing if there aren't wouldn't be any bids
+	if prevTopBidBuilderPubkey == "" {
+		return wasUpdated, ErrFailedUpdatingTopBidNoBids
+	}
+
+	// 4. Don't update if the current bid is the same as the previously known winning bid
+	if payload.Value().Cmp(prevTopBidValue) < 1 {
+		return wasUpdated, nil
 	}
 
-	// Get the actual bid
-	keyBid := r.keyBlockBuilderLatestBids(slot, parentHash, proposerPubkey)
-	bidStr, err := r.client.HGet(context.Background(), keyBid, topBidBuilderPubkey).Result()
+	// Time to update the top bid
+	// 1. Save the execution payload
+	err = r.SaveExecutionPayload(payload.Slot(), payload.ProposerPubkey(), payload.BlockHash(), getPayloadResponse)
 	if err != nil {
-		return err
+		return wasUpdated, err
+	}
+
+	// 2. Save the new top bid - TODO: use transaction
+	err = r.SaveLatestBuilderBid(payload.Slot(), payload.BuilderPubkey().String(), payload.ParentHash(), payload.ProposerPubkey(), reqReceivedAt, getHeaderResponse)
+	if err != nil {
+		return wasUpdated, err
+	}
+
+	// 3. Get the actual bid
+	keyBid := r.keyBlockBuilderLatestBids(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+	bidStr, err := r.client.HGet(context.Background(), keyBid, prevTopBidBuilderPubkey).Result()
+	if err != nil {
+		return wasUpdated, err
+	}
+
+	// 2. Save the top bid - TODO: consider improving by using redis COPY command (if it wouldn't be a hash)
+	keyTopBid := r.keyCacheGetHeaderResponse(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+	err = r.client.Set(context.Background(), keyTopBid, bidStr, expiryBidCache).Err()
+	if err != nil {
+		return wasUpdated, err
 	}
+	wasUpdated = true
 
-	// Save the top bid
-	keyTopBid := r.keyCacheGetHeaderResponse(slot, parentHash, proposerPubkey)
-	return r.client.Set(context.Background(), keyTopBid, bidStr, expiryBidCache).Err()
+	// Finally, update the top bid value
+	keyTopBidValue := r.keyTopBidValue(payload.Slot(), payload.ParentHash(), payload.ProposerPubkey())
+	return wasUpdated, r.client.Set(context.Background(), keyTopBidValue, payload.Value().String(), expiryBidCache).Err()
```

Comment:
Could what's in `keyTopBidValue` change in the meanwhile, creating a race? Same for `keyTopBid`.
Shouldn't the update be a transaction too?

Reply:
yea should be in a transaction too

Reply:
added a comment